drivers/serial/uart_gecko.c

```diff
@@ -301,33 +432,105 @@ static const struct uart_driver_api uart_gecko_driver_api = {
 #endif
 };
 
-#ifdef DT_INST_0_SILABS_GECKO_UART
+/**
+ * @brief Helper to define a single UART/USART pin
+ *
+ * Defines for UART/USART pins are generated from the devicetree
+ * Notation as struct `soc_gpio_pin`, i.e. {port, pin, mode, out}
+ *
+ * @param uart_type UART or USART
+ * @param nr number of UART/USART (0,...,3 for USART0,...,USART3)
+ * @param pin name of pin to define, i.e. RX, TX, RTS, CTS
+ * @param pin_mode GPIO mode of the pin (termination/direction)
+ */
+#define UART_PIN_INIT(uart_type, nr, pin, pin_mode) \
+	{DT_INST_##nr##_SILABS_GECKO_##uart_type##_LOCATION_##pin##_1, \
+	DT_INST_##nr##_SILABS_GECKO_##uart_type##_LOCATION_##pin##_2, pin_mode, 1}
 
-#define PIN_UART0_RXD {DT_INST_0_SILABS_GECKO_UART_LOCATION_RX_1, \
-		DT_INST_0_SILABS_GECKO_UART_LOCATION_RX_2, gpioModeInput, 1}
-#define PIN_UART0_TXD {DT_INST_0_SILABS_GECKO_UART_LOCATION_TX_1, \
-		DT_INST_0_SILABS_GECKO_UART_LOCATION_TX_2, gpioModePushPull, 1}
+/**
+ * @brief Set GPIO assignment for UART pins (RX, TX, RTS, CTS)
+ *
+ * @param uart_type UART or USART
+ * @param nr number of UART/USART (0,...,3 for USART0,...,USART3)
+ */
+#define UART_PINS_WITH_FLOW_CONTROL(uart_type, nr) \
+	.pin_rx  = UART_PIN_INIT(uart_type, nr, RX, gpioModeInput), \
+	.pin_tx  = UART_PIN_INIT(uart_type, nr, TX, gpioModePushPull), \
+	.pin_rts = UART_PIN_INIT(uart_type, nr, RTS, gpioModePushPull), \
+	.pin_cts = UART_PIN_INIT(uart_type, nr, CTS, gpioModeInput),
+
+/**
+ * @brief Set GPIO assignment for UART pins (RX, TX)
+ *
+ * @param uart_type UART or USART
+ * @param nr number of UART/USART (0,...,3 for USART0,...,USART3)
+ */
+#define UART_PINS(uart_type, nr) \
+	.pin_rx  = UART_PIN_INIT(uart_type, nr, RX, gpioModeInput), \
+	.pin_tx  = UART_PIN_INIT(uart_type, nr, TX, gpioModePushPull),
```

Comment:
I believe the driver instantiation code would be simpler if we defined two macros `UART_RX_TX_PINS` and `UART_CTS_RTS_PINS` and then included the latter only if HW flow control is enabled for the instance. It should also make it easier to update the driver in the future to use macro based driver instantiation, similar to how other drivers, e.g. gpio_gecko.c are doing. Though in case of the uart driver this is not going to be straightforward.